### PR TITLE
Allow lazier configuration of jdks

### PR DESCRIPTION
**Pull request by CRogers**
## Before this PR
When configuring an internal Gradle plugin to configure this plugin I ran into a problem - I couldn't work out which JDKs I had without doing an expensive network task - basically what @carterkozak warning me about in the baseline code.

## After this PR
==COMMIT_MSG==
Allow lazy configuration of JDKs
==COMMIT_MSG==

Extra methods allow me to provide all the jdk configuration lazily when it is requested.

### CODE REVIEW

File: gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+import org.gradle.api.Action;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
+
+final class DelayedConfigurableMap<K, V> {
```
Comment (CRogers) [inline gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java]: I tried to use some of the built in `NamedDomainObjectCollection`/`MapProperty` but they really did not seem to do exactly what I wanted - so I build my own thing here.

File: gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    private final ProviderFactory providerFactory;
+    private final Supplier<V> valueFactory;
+    private final List<Provider<Map<K, Action<V>>>> configurations = new ArrayList<>();
```
Comment (carterkozak) [inline gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java]: `>>>>` looks a lot like a merge conflict!

File: gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+import org.gradle.api.Action;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
+
+final class DelayedConfigurableMap<K, V> {
```
Comment (carterkozak) [inline gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java]: What would it look like to abandon the map concept entirely, in favor of something shaped more like a function? I suspect this isn't the only place the map property may bite us

File: gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java (hunk 2 above)
Comment (CRogers) [reply gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java]: This is how you tell it's good code

File: gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java (hunk 3 above)
Comment (CRogers) [reply gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java]: I agree we should probably do that, started it here: https://github.com/palantir/gradle-baseline/pull/2263

Comment (CRogers) [reply gradle-jdks/src/main/java/com/palantir/gradle/jdks/DelayedConfigurableMap.java]: I updated this PR to show what it would look like.

**Pull request merged**